hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/client/SparkRDDWriteClient.java

```diff
@@ -442,8 +443,11 @@ private void updateTableMetadata(HoodieTable<T, JavaRDD<HoodieRecord<T>>, JavaRD
             metaClient, config, context, SparkUpgradeDowngradeHelper.getInstance())
             .run(HoodieTableVersion.current(), instantTime);
         metaClient.reloadActiveTimeline();
-        initializeMetadataTable(Option.of(instantTime));
       }
+      // Initialize Metadata Table to make sure it's bootstrapped _before_ the operation,
+      // if it didn't exist before
+      // See https://issues.apache.org/jira/browse/HUDI-3343 for more details
+      initializeMetadataTable(Option.of(instantTime));
```

Comment:
This is addressing HUDI-3343

Reply:
I get the fix here. But lets say there is a failed commit just before this current writer. Wouldn't the MDT bootstrap also include the files from the failed commit ?  
Or are we gonna rely on MDT bootstrap will not happen if there is any pending operation in data table. And so unless the rollback of the failed commit completes, we may not trigger MDT bootstrap.

Reply:
So the thinking is like following: 
1. MT should be aware of all the (relevant) files in the directory
2. MT should not be bootstrapped in the middle of the operation (to avoid it seeing intermediate state) 

In case of multi-writers -- there're multiple possible scenarios
1. Writers running concurrently: this should be handled by taking the mutex, to make sure MT is bootstrapped once
2. Writer bootstrapping MT running after previous Writer failed to commit (leaving intermediate state in the directory): in that case i'd still argue it's better to MT to be aware of this files than not, so that it's inline with FS-based implementation. Whether these files will be part of the output should be decided by the record-reader, not MT.

Reply:
based on offline discussion, we are good here. MDT intitalization may not kick in if there are any pending operations in data table